Allow JITServer AOT cache stores and loads without a local SCC

cjjdespres opened the pull request:
These changes finish resolving https://github.com/eclipse-openj9/openj9/issues/16721 and allow clients of a JITServer to request JITServer AOT cache load or stores regardless of whether or not a client has a local SCC available. The new implementation (which bypasses the local SCC entirely) is controlled by a new option `-XX:[+|-]JITServerAOTCacheIgnoreLocalSCC`, which is off by default.

For a high level overview of the changes in their entirety, see https://github.com/eclipse-openj9/openj9/issues/16721#issuecomment-1921857539. In a future PR I will expand that comment into a description of the JITServer AOT cache for the `doc/` folder.

cjjdespres commented:
Attn @mpirvu.

mpirvu commented:
> Indeed, clients requesting JITServer AOT loads for particular methods will ignore their local SCC entirely during these requests.

Is this only temporary? Loading from local AOT is faster than loading from the server.

cjjdespres commented:
I meant that they would ignore it while attempting the JITServer AOT load and relocation. The decision to load from the local SCC happens before all of this, I think, and I shouldn't need to change that.

cjjdespres commented on runtime/bcutil/ROMClassBuilder.cpp:
```diff
@@ -864,6 +864,13 @@ ROMClassBuilder::prepareAndLaydown( BufferManager *bufferManager, ClassFileParse
+	 * ROMClasses.
+	 */
+	if (context->isJITServerAOTCacheRequested()) {
+		fixReturnBytecodes(_portLibrary, (J9ROMClass *)romClassBuffer);
```
@pshipton It was suggested that you might know more about `fixReturnBytecodes`. Here I want to use it even when a local SCC is not available. Is this safe to do, or is `fixReturnBytecodes` not applied to non-shared classes for reasons other than efficiency?

pshipton replied:
I'm a bit fuzzy on the details. I think it happens anyway but when storing a class into the shared cache it needs to happen up front because the classes become read-only. @gacholio or @TobiAjila ?

pshipton replied:
Calling fixReturnBytecodes() does have the cost of evaluating all the bytecodes to figure out what the return type should be. I think normally it just modifies it on the fly.

gacholio replied:
Classes in the SCC are read-only, so they must have fixReturnBytecodes done before storing. It's safe to call it when classes are not going into the SCC, but it's unnecessary - the return bytecodes are fixed up on the fly (we assume the bytecodes are "correct" due to verification).

cjjdespres commented:
Just to review the changes needed in the deserializer (which I'll put in the design document too):

Currently, when serving a cached JITServer AOT method, the server will send the method itself (including all of the normal relocation records generated for it) and all the JITServer AOT cache records associated with the method. These JITServer AOT cache records are effectively used to "relocate the relocation records" (as Irwin put it). After the client receives this data, the deserializer at the client will:

1. Take JITServer AOT cache records and look up the entities that they refer to. Thus a `ClassSerializationRecord` will be used to find a `J9Class`, a `ClassLoaderSerializationRecord` will be used to find a `J9ClassLoader`, and so on. This step contains all of the JITServer AOT cache record validation checks, making sure that, e.g., the `J9Class` has a `ROMClass` hash and loader that match the server's records.
2. Find (or store) those entities in the local SCC and retrieve their offsets. These offsets are effectively opaque `uintptr_t` keys that identify these entities within the local SCC.
3. Update the normal relocation records sent along with the cached method so that the offsets they contain are the ones retrieved in (2). This is the relocation-relocation part. (Nothing else in the relocation records needs to be relocated).

After this, the relocation runtime can relocate the method as usual, as the offsets in the record will now be valid with respect to the local SCC.

(A side note at this point: Alexey pointed out that we could just use the JITServer AOT cache record ids as offsets when compiling a method at the server for storage in the JITServer AOT cache. This would let us skip step (3) above in conjunction with the deserializer changes I talk about below, once I get to modifying the code to allow for fresh JITServer AOT cache compilations without a local SCC.)

What we instead need to do in the deserializer is:

1. Repeat step 1 above
2. Assign these entities "deserializer offsets" that will be valid for the duration of a client-server connection, and will not be valid with respect to the local SCC. I think the `idAndType` of the JITServer AOT cache records themselves can be used as offsets at this point, anticipating the future changes to fresh JITServer AOT cache compilations. Previous suggestions included having the offsets come from some incrementing counter, or using pointers to the resolved entities directly as their offsets.
3. Repeat step 3 with these assigned offsets

Once that is done, the relocation work can proceed, which will involve creating a class derived from `TR_J9SharedCache` that we can pass to the relocation runtime. This shared cache interface will consult the deserializer and so be able to convert between the deserializer offsets and normal runtime entities.

cjjdespres commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -896,15 +896,15 @@ ClientSessionData::getClassRecord(J9Class *clazz, JITServer::ServerStream *strea
-
-      if (offset)
+      auto *vmInfo = TR::compInfoPT->getClientData()->getOrCacheVMInfo(stream);
+      if (offset || (aotCacheLoad && !vmInfo->_hasSharedClassCache))
```
This `aotCacheLoad` parameter that I use here should only be temporary - when we ignore the local SCC during fresh JITServer AOT cache compilations this check will need to be updated, because there will be no local offset to store.

cjjdespres commented:
One problem that I've observed with my solution to the ROM classes differing is that if a local SCC is created without `-XX:+JITServerUseAOTCache` the ROM classes in it will be represented differently than the no-SCC case or the case in which we create the local SCC while using the JITServer AOT cache. That's because my changes only take effect if we're a client intending to use the JITServer AOT cache. Thus ROMClasses loaded from the SCC will still differ in hash, and we will get JITServer AOT cache misses for any class that was loaded from the local SCC.

I haven't checked exactly how they differ, but I assume it's the debug information again. This isn't immediately a problem, I suppose - the client just has to run with a local SCC that was created with some option that forces a consistent ROM class representation (as `-XX:+JITServerUseAOTCache` now does) if it wants to use existing cached methods. Of course that's still an annoying requirement.

If the debug info being inlined or not is irrelevant for correctness, then we could solve this by convert ROM classes to a known, consistent representation during comparison so we don't have to deal with this issue. There's already `JITServerHelpers::packROMClass` that does this to an extent, by removing intermediate class data and adding interned strings to the end of the ROM class. Maybe the debug info could be moved to a consistent location in the ROM class at this point as well? 

If the debug info being inlined or not is relevant (and so we don't want ROM classes that differ only in this respect to be equivalent to each other) then we either have to tolerate this problem or find a way of allowing debug info to be stored out-of-line when no local SCC exists. We would do this so that the ROM classes would still have a uniform representation, just one with debug info that's always out of line when using the AOT cache (instead of always inline as I have it in the draft PR). I did see in the code a comment mentioning that this can be done "when the allocation strategy [of the `ROMClassCreationContext`] permits it", so that might be a possibility as well.

I should note that `packROMClass` does not inline any out-of-line debug info for comparison at the moment; it merely zeroes out all SRPs associated with such debug info. At the very least that suggests that the actual content of the debug info is irrelevant for correctness.

cjjdespres commented:
Force pushed to rebase onto master to incorporate the changes in #18344. I also changed a few things so that the PR works when a `readonly` cache is present:

- The class loader table now tracks the names of first-loaded classes even if a chain cannot be found - before it would only track names without chains if a local SCC didn't exist (commit e0aaad9f2abc71d2670f8688c930be67821f5b28)
- AOT loads will be requested from the server even if a particular method couldn't be compiled locally with AOT (commit a715e17ba56ed3396963616cded7aabc24ba340e)
- I simplified a check in `getClassRecord` and removed the `aotCacheLoad` tracking that I had previously introduced - after looking into it I think what I had before was unnecessary (commit 8645b46268b03076260cbff2e6aba02dbd53ed88)

cjjdespres commented:
Force pushed to remove one instance of `aotCacheLoad` I had forgotten about.

cjjdespres commented on runtime/compiler/runtime/JITServerAOTDeserializer.hpp:
```diff
@@ -153,13 +152,14 @@ class JITServerAOTDeserializer
    TR::Monitor *const _methodMonitor;
 
-   PersistentUnorderedMap<uintptr_t/*ID*/, uintptr_t/*SCC offset*/> _classChainMap;
+   PersistentUnorderedMap<uintptr_t/*ID*/, uintptr_t * /*deserializer chain*/> _classChainMap;
```
These chains (and the ones in the well-known classes map) could just be vectors, I suppose.

cjjdespres commented:
I've pushed the deserializer changes, and the start of the JITServer AOT cache documentation in `doc/`.

@AlexeyKhrabrov The actual AOT deserializer changes were more or less as discussed in the related issue, including (perhaps temporarily) removing support for reloading of classes and such.

cjjdespres commented on runtime/compiler/runtime/JITServerAOTDeserializer.cpp:
```diff
@@ -512,74 +571,57 @@ JITServerAOTDeserializer::cacheRecord(const ClassChainSerializationRecord *recor
+      throw std::bad_alloc();
+
+   deserializerChain[0] = chainSize;
+   memcpy(deserializerChain + 1, record->list().ids(), record->list().length());
```
This is incorrect, incidentally. The `idAndType` should be written to the chain, not the IDs themselves.

cjjdespres commented:
The latest changes implement the alternate SCC frontend that ignores the local SCC (if present) and instead looks up the information cached by the deserializer in response to particular queries.

The last thing that potentially needs to be done before this is ready is to deal with the `ROMClass` problem that I mentioned in [this comment](https://github.com/eclipse-openj9/openj9/pull/18301#issuecomment-1802244178).

cjjdespres commented:
There is one other thing - a deserializer reset (which happens when a client disconnects from a server) can happen concurrently. We have to make sure that such a reset cannot occur during relocation, or we have to detect during relocation that such a reset happened and abort relocation if it did.

cjjdespres commented:
I'm not especially happy with `AOTCache.md`, though it is preliminary. What I want to convey is something like:

1. The JITServer AOT cache is a reimplementation of what the local SCC tries to do, which is to use its records to identify RAM classes and RAM methods across JVM invocations, and to encode this information in offsets (opaque `uintptr_t` keys valid only with respect to a particular local SCC) inside a method's relocation records.  The relocation runtime can then use these offsets to materialize the relevant entities during relocation (if they can be found in the current JVM).
2. The old deserializer implementation does all the work of materializing matching RAM classes and methods, but then uses them to store appropriate information (ROM classes, local SCC class chains, etc.) in the local SCC and update the offsets in the method's relocation records to refer to local SCC offsets. This allows the relocation runtime to find the information it needs locally.
3. The new deserializer implementation doesn't touch the local SCC. Instead, it caches the entities it materializes, and updates the offsets in a method's relocation records with its own "deserializer offsets". We then pass in a new `TR_J9SharedCache` subclass during relocation of these methods that ignores the local SCC, and instead uses those offsets to look up the cached entities in the deserializer directly.

AlexeyKhrabrov reviewed (commented):
Not a proper review, just some notes and a few things I noticed in the code.

> The actual AOT deserializer changes were more or less as discussed in the related issue, including (perhaps temporarily) removing support for reloading of classes and such.

I haven't had a chance to look at the code in detail (and won't have time for it for a while). At a high level, how does the new implementation behave when a class is unloaded and later loaded again? By the way, I've seen applications (e.g. some workloads in the Renaissance benchmark suite) where such class re-loading does seem to happen (although I'm not sure how much of the overall performance the affected methods are responsible for).

> There is one other thing - a deserializer reset (which happens when a client disconnects from a server) can happen concurrently. We have to make sure that such a reset cannot occur during relocation, or we have to detect during relocation that such a reset happened and abort relocation if it did.

This existing mechanism and its API turned out to be rather awkward when building other things on top of the deserializer (in my current work on early JIT compilation). This PR might be a good opportunity to get rid of it rather than support it in new code. One likely better mechanism I can think of is a reader-writer lock where threads performing AOT deserialization+load are readers and threads doing a reset are writers. Although I haven't figured the details. If we do keep the current reset detection mechanism, throwing an exception instead of propagating the `wasReset` flag would probably be a better API for it.

cjjdespres commented:
Force-pushed to bring in the changes from https://github.com/eclipse-openj9/openj9/pull/18585 and https://github.com/eclipse-openj9/openj9/pull/18568. I also added a commit that strips the debug info from ROM classes and fixes the bytecodes during ROM class packing, so I was able to drop the changes to the VM and ROM class building that I had before. Stripping the debug info out was one solution to the problem I mentioned in https://github.com/eclipse-openj9/openj9/pull/18301#issuecomment-1802244178. The implementation ended up being a little trickier than I anticipated.

@AlexeyKhrabrov I'll be getting to your initial remarks next. About the class unloading and loading - the new implementation still marks the class as unloaded in the `_classIdMap` (as happens with the current implementation). It just doesn't attempt to find a new matching version of the class if we encounter a cached class entry during deserialization that's been marked as unloaded. The same thing happens with invalidated class loaders and methods.

Moving to a reader-writer lock system sounds fine as well. It wouldn't be hard to write one myself, but is there an implementation already in openj9?

cjjdespres commented on runtime/vm/jvminit.c:
```diff
@@ -2124,8 +2136,14 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
-
-
+#if (defined(J9VM_ARCH_X86) || defined(J9VM_ARCH_S390) || defined(J9VM_ARCH_POWER) || defined(J9VM_ARCH_AARCH64)) && defined(J9VM_OPT_JITSERVER)
+				// Regardless of whether or not the local SCC is enabled, we still need to enable portable AOT if
```
I think this change was @mpirvu's suggestion. It was in the context of a discussion we were having about the `TR_AOTHeader` differing depending on whether or not `-Xshareclasses` was enabled or not.

AlexeyKhrabrov replied:
This needs to be fixed eventually, it's a performance degradation. I remember seeing measurable reductions in throughput with portable AOT back when I worked on the original AOT cache implementation. There is also no point in enabling portable AOT by default at all when running in a container without local SCC (but with AOT cache). Could you please open an issue for this?

I still don't understand what could be different in the `TR_AOTHeader` depending on whether local SCC is enabled. Maybe there is something subtle going wrong with either `featureFlags` or `processDescription`. Have you tried testing this both inside and outside a container, and with portable AOT explicitly enabled or disabled?

cjjdespres replied:
Initially I had individual overrides in particular places, because having/not having an SCC would change some VM properties at startup. Those solved all the header incompatibilities. Explicitly setting the relocatable processor/CPU description in `rossa.cpp` was one of them. Changing how the compressed pointer shift amount was determined was another, as that changes if we're in containers and using an SCC (from what I remember).

Setting the portable AOT flag happened to fix all those individual issues.

cjjdespres replied:
I'll open up an issue

cjjdespres commented:
One thing I've run into in the other half of this issue (code generation at the server) is the use of SCC methods like `isROMClassInSharedCache(romClass, offsetOutputPtr)` and `offsetInSharedCacheFromROMClass(romClass)`, which expect to be able to look up ROM class data in the SCC given just a `J9ROMClass *romClass` input. During a compilation intended for storage in the AOT cache, the offset will, after my changes, correspond to the `idAndType` of the server's class record for the ROM class. We have two ways of getting this record at the server, more or less:

1. The class record map in the AOT cache, with keys being the pair `{J9ROMClass *, AOTCacheClassLoaderRecord *}`
2. The client session data, with keys being (client) `J9Class *` pointers.

Both sources require more information than simply a `J9ROMClass *`. I'm fairly sure wherever these functions are used at the server we always have more information on hand than merely a ROM class. Generally when using these kinds of SCC functions we first look up the ROM class associated to a full `J9Class *`, and then give the resulting `J9ROMClass *` to these methods. So, I could add functions like `offsetInSharedCacheFromClass(clazz)` to the SCC (and other classes, like `J9::AheadOfTimeCompile`) that allow us to communicate this extra information to the actual source of the data. The local SCC can just ignore the extra information, and the server can use it.

A somewhat less invasive change would be to cache some kind of `J9ROMClass * -> AOTCacheClassRecord *` association. I think that runs into the problem that there could be two loaders with distinct `AOTCacheClassLoaderRecord`s that are responsible for loading distinct RAM classes that have identical underlying ROM classes. The current caches do not have this problem.

-------------------

This sort of problem will probably arise in the other SCC lookup methods - the server caches generally want more information than what is typically given to these methods. The `offsetInSharedCacheFromROMMethod` function is another example - the local SCC only needs a `ROMClass *` to get an offset, whereas the server needs a `J9Method *` and its defining `J9Class *`, both of which seem to be available in the only place `offsetInSharedCacheFromROMMethod` is used.

cjjdespres commented:
Force-pushed to address some of Alexey's comments and to bring in https://github.com/eclipse-openj9/openj9/pull/18792.

cjjdespres commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -13181,15 +13181,17 @@ bool TR::CompilationInfo::canProcessJProfilingRequest()
       return false;
 
-#if defined(J9VM_OPT_JITSERVER)
-   if (comp->isDeserializedAOTMethod() && comp->getPersistentInfo()->getJITServerAOTCacheDelayMethodRelocation())
```
I've removed it.

cjjdespres replied:
Just to update this - I haven't removed the option in this final version of the PR, because we still have the option of running with the old implementation.

cjjdespres commented on runtime/compiler/env/J9SharedCache.cpp:
```diff
@@ -897,11 +907,11 @@ TR_J9SharedCache::rememberClass(J9Class *clazz, const AOTCacheClassChainRecord *
-   int32_t numSuperclasses = TR::Compiler->cls.classDepthOf(fe()->convertClassPtrToClassOffset(clazz));
-   int32_t numInterfaces = numInterfacesImplemented(clazz);
+   int32_t numSuperclasses = fe()->numSuperclasses(clazz);
+   int32_t numInterfaces = fe()->numInterfacesImplemented(clazz);
```
Since `numSuperclasses` and `numInterfaces` are used a little later on to fill in the class chain, I got rid of the `necessaryClassChainLength` call instead.

mpirvu reviewed (commented):
I had a bunch of comments, but didn't push them through. I am doing it now, even though I understand the code may be a bit stale

cjjdespres commented:
Pushed to update the PR given the changes that have already been merged. The last changes ended up touching more files than I'd anticipated because I had to override a few more things (thunks, some relocatable method handling functions) during an AOT cache store.

I've left out the JITServer AOT cache documentation file for now. I think I'll end up expanding https://github.com/eclipse-openj9/openj9/issues/16721#issuecomment-1921857539 as a replacement for the old one that used to be in this PR.

cjjdespres commented:
Force-pushed to bring in https://github.com/eclipse-openj9/openj9/pull/18971

mpirvu commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -3105,13 +3106,24 @@ remoteCompile(J9VMThread *vmThread, TR::Compilation *compiler, TR_ResolvedMethod
+   bool aotCacheLoad = false;
+   if (persistentInfo->getJITServerAOTCacheIgnoreLocalSCC())
+      {
+      aotCacheStore = entry->_useAotCacheCompilation &&
```
Why isn't `persistentInfo->getJITServerUseAOTCache()` included here? Is it guaranteed by one of the two remaining parts?

cjjdespres replied:
I put the test in `preCompilationTasks` here:

https://github.com/eclipse-openj9/openj9/pull/18301/files#diff-6b9288dd0884d4a19c6bb2f68e31dea915c9acf93d3e25bedeb9fa0c1719d738R7936-R7943

when setting `eligibleForRemoteAOTNoSCCCompile`, so if `entry->_useAotCacheCompilation` is true we should know that we're running with `getJITServerUseAOTCache()`.

mpirvu commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -3105,13 +3106,24 @@ remoteCompile(J9VMThread *vmThread, TR::Compilation *compiler, TR_ResolvedMethod
+      {
+      aotCacheStore = useAotCompilation && persistentInfo->getJITServerUseAOTCache() &&
+                      compInfo->methodCanBeJITServerAOTCacheStored(compiler->signature(), compilee->convertToMethod()->methodType());
+      aotCacheLoad = aotCacheStore && !entry->_doNotLoadFromJITServerAOTCache &&
```
Previously, AOT cache loads could be done only on interpreted methods. Here the check on `!TR::CompilationInfo::isCompiled(method)` has disappeared. Was this done on purpose?

cjjdespres replied:
Sorry, this is an artifact of the commit structure - if you look at the diff overall there wasn't a `!TR::CompilationInfo::isCompiled(method)` test there to begin with. (I added it in the "change AOT load criteria commit"). A method not being compiled is a prerequisite for  `entry->_useAotCompilation`, so it would be redundant to add when `getJITServerAOTCacheIgnoreLocalSCC()` is false.

It does need to be added for determining `aotCacheLoad` when `getJITServerAOTCacheIgnoreLocalSCC()` is true, though. I took the criteria from the existing ones for AOT loads in `precompilationTasks()` here:

https://github.com/eclipse-openj9/openj9/blob/6a6d7afd73759f7bf4c786055bcb02bb35343030/runtime/compiler/control/CompilationThread.cpp#L7707-L7712

omitting the ones that didn't seem relevant. Though, now that I look at that code again I see that there's a `!_jitConfig->inlineFieldWatches` component that I might want to add, because I put that in the eligibility criteria for AOT cache stores.

mpirvu reviewed (commented):
LGTM. I have small suggestions/questions here and there

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit,alinux64jit jdk17

mpirvu commented:
A point Alexey raised a while ago and didn't have time to address it:
Here 
https://github.com/eclipse-openj9/openj9/blob/70e835eb2ee6d4ae5c975646d836dd2506d971c8/runtime/compiler/control/JITClientCompilationThread.cpp#L189-L191
we must capture all messages that can end a compilation. `MessageType::AOTCache_serializedAOTMethod` is missing from the check and you added a new one as well that must be captured: `MessageType::AOTCache_storedAOTMethod`

cjjdespres commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -11339,7 +11339,9 @@ void TR::CompilationInfoPerThreadBase::logCompilationSuccess(
                TR_VerboseLog::write(" remote");
-               if (compiler->isDeserializedAOTMethod())
+               if (compiler->isDeserializedAOTMethodStore())
+                  TR_VerboseLog::write(" aotStored");
```
Changed. I also moved `aotStored` to the end, because I think one of the tests looks for `remote deserialized` specifically.

cjjdespres commented:
I think that should be everything. Of course, https://github.com/eclipse-openj9/openj9/pull/18976 isn't in yet, so depending on what tests are run that error will show up. I can rebase once that's in.

cjjdespres commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -3181,24 +3184,27 @@ remoteCompile(J9VMThread *vmThread, TR::Compilation *compiler, TR_ResolvedMethod
       {
-      classChainOffset = compiler->fej9vm()->sharedCache()->rememberClass(clazz);
-      if (TR_SharedCache::INVALID_CLASS_CHAIN_OFFSET != classChainOffset)
+      // TODO: this check should omit the aotCacheStore part when we support ignoring the local SCC during AOT cache stores.
```
It's done in this one, sorry. This was from a commit that was in the original PR, and I forgot to remove it in the original commit. Instead I removed the `aotCacheStore` part of the condition in the later commit "Enable new JITServer AOT cache store implementation" (currently 5bbed93340b543cbfe72bf1f0f722067990601d3). It's the same thing I did in a couple of other things you highlighted.

cjjdespres commented:
Just to record it here - the test failures are likely related to the PR that https://github.com/eclipse-openj9/openj9/pull/18981 reverted. Once this PR is rebased to bring in that revert and the ROM class walk fix, we should be good to try PR testing again.

cjjdespres commented:
Force-pushed to rebase onto master.

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit,alinux64jit jdk17

mpirvu commented:
zlinux failed the `cmdLineTester_criu_jitserverPostRestore_1` with "can't bind server address: Address already in use".
This is a known issue.

keithc-ca commented on runtime/compiler/env/ClassLoaderTable.cpp:
```diff
@@ -225,26 +225,18 @@ TR_PersistentClassLoaderTable::associateClassLoaderWithClass(J9VMThread *vmThrea
+   // If we are using the JITServer AOT cache and ignoring the local SCC, we need to remember the name of clazz
+   // with or without chain. Otherwise (not using AOT cache or not ignoring the local SCC) there is no point in continuing
+   // without a chain.
+   if (!chain && (!useAOTCache || !_persistentMemory->getPersistentInfo()->getJITServerAOTCacheIgnoreLocalSCC()))
```
This doesn't compile without `defined(J9VM_OPT_JITSERVER)`, e.g. on macOS; see [failed build](https://openj9-jenkins.osuosl.org/job/Build_JDK22_aarch64_mac_Personal/9):
```
[2024-02-20T21:51:46.296Z] /Users/jenkins/workspace/Build_JDK22_aarch64_mac_Personal/openj9/runtime/compiler/env/ClassLoaderTable.cpp:239:76: error: no member named 'getJITServerAOTCacheIgnoreLocalSCC' in 'TR::PersistentInfo'
[2024-02-20T21:51:46.296Z]    if (!chain && (!useAOTCache || !_persistentMemory->getPersistentInfo()->getJITServerAOTCacheIgnoreLocalSCC()))
[2024-02-20T21:51:46.296Z]                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
[2024-02-20T21:51:46.296Z] 1 error generated.
[2024-02-20T21:51:46.296Z] make[6]: *** [runtime/compiler/CMakeFiles/j9jit.dir/env/ClassLoaderTable.cpp.o] Error 1
```

mpirvu replied:
I'll put a fix shortly

AlexeyKhrabrov commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -3088,10 +3116,24 @@ remoteCompile(J9VMThread *vmThread, TR::Compilation *compiler, TR_ResolvedMethod
+      {
+      aotCacheStore = entry->_useAOTCacheCompilation &&
+                      compInfo->methodCanBeJITServerAOTCacheStored(compiler->signature(), compilee->convertToMethod()->methodType());
+      aotCacheLoad = persistentInfo->getJITServerUseAOTCache() &&
```
I think this changes the semantics of the `-Xaot:jitserverAOTCacheStoreExclude=` option, which previously implied `LoadExclude` as well (and still implies for the local SCC mode).

AlexeyKhrabrov commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -3305,6 +3350,52 @@ remoteCompile(J9VMThread *vmThread, TR::Compilation *compiler, TR_ResolvedMethod
+         if (aotCacheLoad)
+            deserializer->incNumCacheMisses();
+
+         auto method = SerializedAOTMethod::get(methodStr);
```
Some of the code duplicated in the next branch can be factored out.

AlexeyKhrabrov commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -3585,7 +3678,7 @@ remoteCompile(J9VMThread *vmThread, TR::Compilation *compiler, TR_ResolvedMethod
             }
 
-         if (!compiler->getOption(TR_DisableCHOpts) && !useAotCompilation && !compiler->isDeserializedAOTMethod())
+         if (!compiler->getOption(TR_DisableCHOpts) && !useAotCompilation && (compiler->isDeserializedAOTMethodStore() || !compiler->isDeserializedAOTMethod()))
```
I'm having trouble understanding this condition. Shouldn't `isDeserializedAOTMethodStore()` imply that this is an AOT compilation?

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -7992,7 +8016,33 @@ TR::CompilationInfoPerThreadBase::preCompilationTasks(J9VMThread * vmThread,
+            // TODO: is -Xaot:forceaot recognized when -Xshareclasses:none is specified?
+            canDoRelocatableCompile = true;
+            }
+         else if (!preferLocalComp(entry))
```
Would be nice to factor this logic out (although to be fair, this whole method is a bit of a mess and could use some much heavier refactoring...).

AlexeyKhrabrov reviewed (commented):
I have some concerns about changes in control logic, a few questions, and some suggestions for possible future refactoring.
